File: src/app/browser/header/menu.cpp
#include "menu.hpp"

using namespace app::browser::header;

Menu::Menu(
    const Glib::RefPtr<Gio::SimpleAction> & ACTION__DEBUG,
    const Glib::RefPtr<Gio::SimpleAction> & ACTION__QUIT,
    const Glib::RefPtr<Gio::SimpleAction> & ACTION__SESSION_CLEAN,
    const Glib::RefPtr<Gio::SimpleAction> & ACTION__SESSION_RESTORE,
    const Glib::RefPtr<Gio::SimpleAction> & ACTION__SESSION_SAVE,
    const Glib::RefPtr<Gio::SimpleAction> & ACTION__TAB_APPEND,
    const Glib::RefPtr<Gio::SimpleAction> & ACTION__TAB_PIN,
    const Glib::RefPtr<Gio::SimpleAction> & ACTION__TAB_CLOSE,
    const Glib::RefPtr<Gio::SimpleAction> & ACTION__TAB_CLOSE_ALL,
    const Glib::RefPtr<Gio::SimpleAction> & ACTION__TAB_PAGE_NAVIGATION_HISTORY_BACK,
    const Glib::RefPtr<Gio::SimpleAction> & ACTION__TAB_PAGE_NAVIGATION_HISTORY_FORWARD,
    const Glib::RefPtr<Gio::SimpleAction> & ACTION__TAB_PAGE_NAVIGATION_RELOAD
) {
    // Init widget
    set_tooltip_text(
        _("Menu")
    );

    // Init components @TODO make builder
    const auto MENU = Gio::Menu::create();

        const auto MENU__SESSION = Gio::Menu::create();

            MENU__SESSION->append(
                _("Clean"),
                get_action_detailed_name(
                    ACTION__SESSION_CLEAN
                )
            );

            MENU__SESSION->append(
                _("Restore"),
                get_action_detailed_name(
                    ACTION__SESSION_RESTORE
                )
            );

            MENU__SESSION->append(
                _("Save"),
                get_action_detailed_name(
                    ACTION__SESSION_SAVE
                )
            );

        MENU->append_submenu(
            _("Session"),
            MENU__SESSION
        );

        const auto MENU__TAB = Gio::Menu::create();

            MENU__TAB->append(
                _("New.."),
                get_action_detailed_name(
                    ACTION__TAB_APPEND
                )
            );

            MENU__TAB->append(
                _("Pin"),
                get_action_detailed_name(
                    ACTION__TAB_PIN
                )
            );

                const auto MENU__TAB_PAGE = Gio::Menu::create();

                    const auto MENU__TAB_PAGE_NAVIGATION = Gio::Menu::create();

                        const auto MENU__TAB_PAGE_NAVIGATION_HISTORY = Gio::Menu::create();

                            MENU__TAB_PAGE_NAVIGATION_HISTORY->append(
                                _("Back"),
                                get_action_detailed_name(
                                    ACTION__TAB_PAGE_NAVIGATION_HISTORY_BACK
                                )
                            );

                            MENU__TAB_PAGE_NAVIGATION_HISTORY->append(
                                _("Forward"),
                                get_action_detailed_name(
                                    ACTION__TAB_PAGE_NAVIGATION_HISTORY_FORWARD
                                )
                            );

                        MENU__TAB_PAGE_NAVIGATION->append_submenu(
                            _("History"),
                            MENU__TAB_PAGE_NAVIGATION_HISTORY
                        );

                        MENU__TAB_PAGE_NAVIGATION->append(
                            _("Reload"),
                            get_action_detailed_name(
                                ACTION__TAB_PAGE_NAVIGATION_RELOAD
                            )
                        );

                    MENU__TAB_PAGE->append_submenu(
                        _("Navigation"),
                        MENU__TAB_PAGE_NAVIGATION
                    );

            MENU__TAB->append_submenu(
                _("Page"),
                MENU__TAB_PAGE
            );

            const auto MENU__TAB_CLOSE = Gio::Menu::create();

                MENU__TAB_CLOSE->append(
                    _("Active tab"),
                    get_action_detailed_name(
                        ACTION__TAB_CLOSE
                    )
                );

                // @TODO
                /*
                MENU__TAB_CLOSE->append(
                    _("All tabs to left"),
                    get_action_detailed_name(
                        ACTION__TAB_CLOSE_LEFT
                    )
                );

                MENU__TAB_CLOSE->append(
                    _("All tabs to right"),
                    get_action_detailed_name(
                        ACTION__TAB_CLOSE_RIGHT
                    )
                );
                */
                MENU__TAB_CLOSE->append(
                    _("All tabs"),
                    get_action_detailed_name(
                        ACTION__TAB_CLOSE_ALL
                    )
                );

            MENU__TAB->append_submenu(
                _("Close"),
                MENU__TAB_CLOSE
            );

        MENU->append_submenu(
            _("Tab"),
            MENU__TAB
        );

        const auto MENU__TOOL = Gio::Menu::create();

            MENU__TOOL->append(
                _("Debug"),
                get_action_detailed_name(
                    ACTION__DEBUG
                )
            );

        MENU->append_submenu(
            _("Tool"),
            MENU__TOOL
        );

        MENU->append(
            _("Quit"),
            get_action_detailed_name(
                ACTION__QUIT
            )
        );

    set_menu_model(
        MENU
    );
}

Glib::ustring Menu::get_action_detailed_name(
    const Glib::RefPtr<Gio::SimpleAction> & ACTION
) {
    return Glib::ustring::sprintf(
        "win.%s", ACTION->get_name()
    );
}

File: src/app/browser/header/tray/menu.rs
use gtk::{
    gio::{self, SimpleAction},
    glib::{gformat, GString},
    prelude::ActionExt,
    MenuButton,
};

use std::sync::Arc;

pub struct Menu {
    gobject: MenuButton,
}
#[rustfmt::skip] // @TODO template builder?
impl Menu {
    pub fn new(
        action_tool_debug: Arc<SimpleAction>,
        action_tool_profile_directory: Arc<SimpleAction>,
        action_quit: Arc<SimpleAction>,
        action_tab_append: Arc<SimpleAction>,
        action_tab_close: Arc<SimpleAction>,
        action_tab_close_all: Arc<SimpleAction>,
        action_tab_page_navigation_base: Arc<SimpleAction>,
        action_tab_page_navigation_history_back: Arc<SimpleAction>,
        action_tab_page_navigation_history_forward: Arc<SimpleAction>,
        action_tab_page_navigation_reload: Arc<SimpleAction>,
        action_tab_pin: Arc<SimpleAction>,
    ) -> Self {
        // Init model
        let model = gio::Menu::new();

            let model_tab = gio::Menu::new();
                model_tab.append(Some("New"), Some(&detailed_action_name(action_tab_append)));
                model_tab.append(Some("Pin"), Some(&detailed_action_name(action_tab_pin)));

                let model_tab_page = gio::Menu::new();

                    let model_tab_page_navigation = gio::Menu::new();
                        model_tab_page_navigation.append(Some("Base"), Some(&detailed_action_name(action_tab_page_navigation_base)));

                        let model_tab_page_navigation_history = gio::Menu::new();
                            model_tab_page_navigation_history.append(Some("Back"), Some(&detailed_action_name(action_tab_page_navigation_history_back)));
                            model_tab_page_navigation_history.append(Some("Forward"), Some(&detailed_action_name(action_tab_page_navigation_history_forward)));

                        model_tab_page_navigation.append_submenu(Some("History"), &model_tab_page_navigation_history);
                        model_tab_page_navigation.append(Some("Reload"), Some(&detailed_action_name(action_tab_page_navigation_reload)));
                         // @TODO model_tab_page_navigation.append(Some("Bookmark"), Some("win.tab_page_bookmark"));

                    model_tab_page.append_submenu(Some("Navigation"), &model_tab_page_navigation);

                model_tab.append_submenu(Some("Page"), &model_tab_page);

                let model_tab_close = gio::Menu::new();
                    model_tab_close.append(Some("Current"), Some(&detailed_action_name(action_tab_close)));
                    model_tab_close.append(Some("All"), Some(&detailed_action_name(action_tab_close_all)));

                model_tab.append_submenu(Some("Close"), &model_tab_close);

            model.append_submenu(Some("Tab"), &model_tab);

            let model_tool = gio::Menu::new();
                model_tool.append(Some("Debug"), Some(&detailed_action_name(action_tool_debug)));
                model_tool.append(Some("Profile directory"), Some(&detailed_action_name(action_tool_profile_directory)));

            model.append_submenu(Some("Tool"), &model_tool);

            model.append(Some("Quit"), Some(&detailed_action_name(action_quit)));

        // Init widget
        let gobject = MenuButton::builder().tooltip_text("Menu").build();
        gobject.set_menu_model(Some(&model));

        // Result
        Self { gobject }
    }

    // Getters
    pub fn gobject(&self) -> &MenuButton {
        &self.gobject
    }
}

// Private helpers
fn detailed_action_name(action: Arc<SimpleAction>) -> GString {
    gformat!("win.{}", action.name()) // @TODO find the way to ident parent group
                                      // without application-wide dependencies import
                                      // see also src/app/action.rs
}
